### ISSUE TITLE
Invert dependency between Scalaz 8 IO and Scalaz 8

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz#1798]
Scalaz 8 IO is integrated thoroughly into Scalaz 8, and the effect package depends on the core data types and type classes. However, it's possible and relatively straightforward to flip this dependency around, allowing Scalaz 8 IO to live in its own module, with Scalaz 8 and even Scalaz 7.3 depending on this module. This ticket outlines the steps involved in such a process.

 - [ ] Create a `scalaz-io` repository, or co-opt `ioeffect`.
 - [ ] Refactor Scalaz 8 IO to not use any core data types, instead relying on "church encoding" for things like `\/` and `Maybe`.
 - [ ] Move Scalaz 8 IO instances from `scalaz.effect` into core, and flip the dependency.
 - [ ] Put a code freeze on Scalaz 8 IO, and copy the Scalaz 8 IO code into `scalaz.io` (as per proposed package organization), and into the separate repository for `scalaz-io`.
 - [ ] Setup package publishing for Scalaz 8 IO.
 - [ ] Publish a first version of Scalaz 8 IO.
 - [ ] Delete the Scalaz 8 IO from Scalaz 8, and instead depend on the published `scalaz-io` package.
 - [ ] Delete the `IO` implementation and associated data types from Scalaz 7.3, and instead, depend directly on `scalaz-io`, adding the necessary instances in core.

**Comment by fommil** [comment]
if we co-opt ioeffect we'll need to provide a 7.2 backcompat module with the orphan instances. Seems a good reason to make a 3.0 release.

What about the package name?

I like the thought of having access to IO in core. No excuses for impure code!

**Comment by ktonga** [comment]
I could start with this one. Can someone create the new repo?

**Comment by fommil** [comment]
@ktonga let's just use ioeffect. We can cut a 3.0.0 release.

**Comment by ktonga** [comment]
I liked more `scalaz-io` it aligns better with the naming convention in the rest of the repos :(

**Comment by jdegoes** [comment]
@fommil @ktonga I do prefer to use a new repo.

Ioeffect can stay for 7.2. We should rename it, of course.

**Comment by ktonga** [comment]
Also in 7.2 we have to keep the package with a name other than `effect`
What you think about using `scalaz.io` as the base package in the new repo? it can be used across all versions. `effect` is too generic :)

**Comment by fommil** [comment]
`scalaz.io` wouldn't be good because so many people use `io.` domain names these days and it causes a nightmare when mixed with relative imports.

new repo it is!

**Comment by ktonga** [comment]
fs2 has an `io` sub-package and doesn't seem to be causing any problems. Just sayin'

**Comment by fommil** [comment]
oh, yes it is... so much.... not used it with circe lately? :smile:

**Comment by jdegoes** [comment]
@fommil Aren't they required to use `io.scalaz....`, i.e. reversed domain name? Technically, we should be using `org.scalaz` on everything.

**Comment by ktonga** [comment]
Namespaces are exactly to solve this kind of problems and we are using them to create even more problems? If you have `fs2.io` and `circe.io` just don't use relative packages.

**Comment by fommil** [comment]
that's the thing... it's `io.circe` and there are many others and this causing problems because there is always an `io` package in scope. They are not the only ones, it's just an example.

Imagine trying to call a package `com` or `org` and thing of the problems you'd see.

**Comment by ktonga** [comment]
I could totally call a package `com.domain.com`, the thing is that the package name is not just `com` but the whole thing. Using relative (pre-imported) packages is only there for convenience, and it is not convenient when it causes name clashing :smile:

**Comment by hrhino** [comment]
I don't see what the potential problem is? Is the idea that if you `import scalaz._`, you can't then `import io.circe._`? I dunno how annoying that would prove in practice; you can just put the circe import above the scalaz one.

I agree with @ktonga that `effect` is too generic. 

Either way, I'd rather have the feature than my choice of package name, so I won't argue.

**Comment by fommil** [comment]
yeah, that's pretty much the problem. But it's not just circe, it also means if you have a `import fs2._` around it'll mess things up, and `import scalaz._` will always be getting in the way. Plus a bunch of other libs... I just think `io` is too overloaded.

**Comment by jdegoes** [comment]
We can keep bike-shedding on this, btw. We may come up with a short package name that's not too generic and fits with the existing conventions. I do think in practice package name conflicts will happen rarely in practice and would not happen at all if the generic import were something like `import scalaz.Predef._` instead of `import scalaz._; import Scalaz._`.

**Comment by fommil** [comment]
Why not just use the scalaz namespace?

**Comment by hrhino** [comment]
:+1: to `import scalaz.Prelude._`.

**Comment by jdegoes** [comment]
You mean so `import scalaz._` brings in everything? It'd certainly be nice if there were no name clashes between any types.

**Comment by jdegoes** [comment]
@fommil Oh and why are you still awake? 😉

**Comment by ktonga** [comment]
At least the consensus seems to be that `scalaz.effect` is not the right place for `IO`.

Bikeshedding ends here :relaxed:. Please react to this comment with your preference:

:heart: We go full-pureboi all the way and put `IO`-related stuff in `scalaz`
:+1: We put it in `scalaz.io` and we deal with potential problems later
:-1: We leave it in `scalaz.effect`

Thanks!

**Comment by hrhino** [comment]
Thanks for stealing our paint, @ktonga. :P

**Comment by jdegoes** [comment]
I made my vote for `scalaz.io` but it's contingent on changing the "default import" from `import scalaz._` to something more precise, like `import scalaz.Prelude._`.

If there's ever a reason a user has to import `scalaz._`, then I agree with @fommil this will cause clashes.

Another option is to run with `testz` theme and use `ioz`. In fact this could be used pervasively as a way to make all Scalaz packages unique without extraneous characters or giving up useful, common names like `io`.

**Comment by fommil** [comment]
:+1: for the `z` suffix

**Comment by ktonga** [comment]
Is it possible to assign this one to me and mark it as WIP?

**Comment by jdegoes** [comment]
@ktonga Unfortunately, it is not possible to assign it to you yet. I will add you to the organization and see if that fixes the problem.

**Comment by jdegoes** [comment]
@ktonga Making excellent progress!
